### ISSUE TITLE
mitmproxy does not see traffic from Android app

### ISSUE THREAD

**Issue by bridgewaterrobbie** [mitmproxy/mitmproxy#2054]
##### Steps to reproduce the problem:

1. Setup Mitmproxy for android, setting up the certificate (You can verify you see web traffic)
2. Go to app and try to cause network activity (App uses AWS and cognate)0
Expected result:
App connects via proxy and traffic is shown in mitmproxy

Actual result:
App fails to connect to server- Mitmproxy sees no activity

On iOS this works fine with the same app. 

##### Any other comments? What have you tried so far?

The behavior seems similar to what happens when you try to connect to an https site before installing the certificate- no network traffic is seen until you accept the insecure connection. 

##### System information

Google Pixel running 7.1.1

**Comment by mhils** [comment]
I'm afraid this is the first bug report where we see https://android-developers.googleblog.com/2016/07/changes-to-trusted-certificate.html in action (see also discussion at https://twitter.com/mitmproxy/status/751193333251190784). 

I tried to ask Google folks about this at a couple of places. Sorry for the drastic words, but my preliminary conclusion is that the Android team just doesn't give a shit about your privacy here. I just don't see the threat model they are addressing with this change.

It's a bit of a shame for Google as the Chrome security team does a fantastic job at advancing the state of the TLS ecosystem, but they are apparently not talking to the Android team. ¯\\\_(ツ)_/¯ If you know any Google folks involved in that decision, we'd be very happy to talk and hear their point of view.

**Comment by victorhooi** [comment]
There was a discussion about this earlier on HN:

https://news.ycombinator.com/item?id=12061320

I hit this issue myself this morning, trying to debug an issue in an Android app as well.

The mitmproxy docs (http://docs.mitmproxy.org/en/stable/certinstall.html) mention this in passing:

> If you want to intercept the pinned connections, you need to patch the application manually. For Android and (jailbroken) iOS devices, various tools exist to accomplish this.

But they don't actually mention what these various tools are.

I believe if you have a rooted Android device, and XPosed installed - perhaps you could use:

https://github.com/Fuzion24/JustTrustMe

However, I haven't verified it myself. If somebody is able to verify it - should we reference it in the docs?

Any other approaches people can suggest?

**Comment by victorhooi** [comment]
Hmm Xposed doesn't support Android 7.x =(.

Any other suggestions?

**Comment by bridgewaterrobbie** [comment]
I added the "Trusting custom CAs for debugging" code (With a modification since the XML is wrong), but it did not allow the app to go through the proxy.

Might be notable though, I could use and track a button that spawns a web view, just not the login functionality. This makes me think the issue for me isn't entirely google, it might be partly cognito(Again, my iOS version works fine).

I also think it aught to work if you install the mitm certificate as a system trusted certificate, but am unsure how to go about it (It should be possible when rooted though right?)

**Comment by mhils** [comment]
@chadbrubaker, I see you authored the Android blog post, can you shed some light on the team's motivation for this change in the trust policy? Sorry for the cold mention, just curious what's the reasoning behind all this.

(For the record, the new API looks considerably better, but the hiding-traffic-from-users aspect is very irritating)

**Comment by chadbrubaker** [comment]
The crux of the motivation is in the blog post -- application data is secure by default. Our goal is to protect end users, so while I enjoyed the snarky analysis up thread that's not at all why :). Protecting application data _is_ protecting user data and by extension user privacy at a fundamental level.

The change does add additional work for reverse engineering or testing applications, and I'm sorry for that, but neither of those tasks are in the threat model for this feature. It does not prevent nor attempt to prevent you from doing those kinds of things.

It always been the case that an application could choose not to trust your MiTM CA and there are many things you can do as a reverse engineer in those situations. Patching the TLS code in OS is almost always massive overkill and solving the problem at the wrong level, IMO, but how you want to go about that is up to you.

If you do find apps with security or privacy issues do let us know at security@android.com.

@victorhooi Keep in mind that XPosed module [disables **all** TLS verification](https://github.com/Fuzion24/JustTrustMe/blob/master/app/src/main/java/just/trust/me/Main.java#L309) which is something you should **never** do and makes it trivially MiTMable, don't use that on a device with private data and wipe the device when you're done.

@bridgewaterrobbie debug-overrides are only used when android:debuggable is set. The config under 'Trusting user-added CAs for all secure connections' should work just fine for you. Also sorry about that typo in the blog, oops!

**Comment by mhils** [comment]
@chadbrubaker: Thanks for getting back so quickly - I appreciate it!

> Our goal is to protect end users, so while I enjoyed the snarky analysis up thread that's not at all why :). Protecting application data *is* protecting user data and by extension user privacy at a fundamental level.

What is your threat model regarding user-added CAs? In all other settings I know of, user-added CAs are considered particularly trustworthy. For example, Chrome, Firefox and IE will not enforce HPKP if they encounter a cert from a user-added CA. Why does Android consider user-added CAs less trustworthy than normal CAs, whereas all other major platforms do the contrary?

> The change does add additional work for reverse engineering or testing applications, and I'm sorry for that, but neither of those tasks are in the threat model for this feature. It does not prevent nor attempt to prevent you from doing those kinds of things.

This may not be your intention, but unfortunately I feel it's a practical consequence. You and I know how to circumvent that, but [journalists at the WSJ](http://blogs.wsj.com/digits/2012/12/07/sites-sharing-personal-details-the-journal-methodology/), [information privacy lawyers](http://cyber.jotwell.com/an-internet-x-ray-machine-for-the-masses/) or [the guy who found that Path uploaded your address book ](https://gizmodo.com/5883118/path-uploads-your-entire-address-book-to-find-your-friends) probably won't bother as much. A lot of those issues are just found by people who don't run custom roms. Not providing a developer option for this raises the entry bar by a lot.


Sorry if this comes off harsh - I really don't want to argue against more secure defaults, but I'd appreciate a good explanation what leads your team think that Android should behave different than all other ecosystems here. If there are good reasons, I'd love to hear them here or privately if you'd prefer to take it off-list (email is on my GitHub).

Thanks!
Max

**Comment by bridgewaterrobbie** [comment]
Just throwing out, after using 
```
<network-security-config>  
     <debug-overrides>  
          <trust-anchors>  
               <!-- Trust preinstalled CAs -->  
               <certificates src="system" />  
               <!-- Additionally trust user added CAs -->  
               <certificates src="user" />  
          </trust-anchors>  
     </debug-overrides>  
</network-security-config>
```
in my network security config things started working. Not sure if the system part was needed, but wanted to cover everything so I could get back to work.

**Comment by cortesi** [comment]
@chadbrubaker Any further comment on this issue? I think everyone can agree that the practical impact of this move will be to greatly reduce the amount of scrutiny that apps receive. I strongly feel that the net effect will be negative for users. We're nowhere near the point where the Android ecosystem is well-controlled enough to tilt the balance away from transparency in this way.

**Comment by javabrett** [comment]
Would the option to trust user certificates be a good candidate for adding to Developer Mode preferences? Assuming that dialog is well secured, this should strike a good balance between user privacy and security developer need.

It would be a great shame to lose this ability on Android, or for it to require custom roms.

**Comment by victorhooi** [comment]
Has anybody looked into this, and whether it works?

https://github.com/levyitay/AddSecurityExceptionAndroid

Apparently you can simply pass this an APK file, and it will automate adding the network exceptions into it (i.e. the stuff above in https://github.com/mitmproxy/mitmproxy/issues/2054#issuecomment-282775181) It also requires apktool:

https://ibotpeaches.github.io/Apktool/

**Comment by raymontag** [comment]
@victorhooi Thanks worked like a charm for several apps.

@mhils @cortesi Maybe you should consider to put this as a workaround into the official documentation. My guess is that the situation will not change anytime soon.

**Comment by bogomazov** [comment]
I don't like the idea of recompiling different apks, is there any update in here? I will really appreciate universal solutions for all apps installed on my device.

**Comment by Lekensteyn** [comment]
Ran into this issue as well while trying to check for privacy implications of an app. The [blog post][0] and [docs][4] linked above does explain what happened, but not how. Source code references follows based on nougat-mr2.3.

The change was made in commit https://github.com/android/platform_frameworks_base/commit/32d2a1024f75f7e917f2aca18d34322a46d36bcb, see [frameworks/base:core/java/android/security/net/config/NetworkSecurityConfig.java][1]. After that change, only the [system root store][2] (`/etc/security/cacerts`) is considered.

`NetworkSecurityConfig.getDefaultBuilder` is [used in XmlConfigSource.java][3] when parsing the configuration file, if the option is not found in the APK, then it uses that default (which only includes the system root).

XmlConfigSource (and the fallback to getDefaultBuilder) is used in NetworkSecurityConfig, which is used all over the place. And unlike permissions, this configuration does not seem cached in /data/system/.
___
Editing each APK after installation and upgrades seem fragile...

So let's turn to the alternative since I have root, adding the certificate to the root store in `/etc/security/cacerts/`. Certificates here are stored based on the hash of the Subject field plus an index number (increasing from 0; in case the same certificate is issues with multiple hash algorithms, e.g. SHA-1 and SHA-256).

```
$ openssl x509 -noout -subject_hash_old -in ~/.mitmproxy/mitmproxy-ca-cert.pem
c8750f0d
```
Now you need to copy your PEM-encoded CA file to the system partition, like:
```
$ adb root
$ adb push ~/.mitmproxy/mitmproxy-ca-cert.pem /etc/security/cacerts/c8750f0d.0
$ adb shell chmod 644 /etc/security/cacerts/c8750f0d.0
```
In my case, my `/system` was readonly and I could not write it. So as workaround, I cloned the cacerts directory from system to userdata, added the cert and bind-mounted it over the system directory.
```
ca=~/.mitmproxy/mitmproxy-ca-cert.pem
hash=$(openssl x509 -noout -subject_hash_old -in $ca)
adb root
adb shell cp -a /etc/security/cacerts /data/
adb push $ca /data/cacerts/$hash.0
adb shell chmod 644 /data/cacerts/$hash.0
adb shell mount --bind /data/cacerts /etc/security/cacerts
adb unroot  # optional: restart adbd as normal user
```

Restarted the app, and it works! No more "Trust anchor for certification path not found." errors!

Note: this needs to be done once every time you restart the device and want to perform a mitm.

 [0]: https://android-developers.googleblog.com/2016/07/changes-to-trusted-certificate.html
 [1]: https://github.com/android/platform_frameworks_base/blob/nougat-mr2.3-release/core/java/android/security/net/config/NetworkSecurityConfig.java#L161
 [2]: https://github.com/android/platform_frameworks_base/blob/nougat-mr2.3-release/core/java/android/security/net/config/SystemCertificateSource.java
 [3]: https://github.com/android/platform_frameworks_base/blob/nougat-mr2.3-release/core/java/android/security/net/config/XmlConfigSource.java#L357
 [4]: https://developer.android.com/training/articles/security-config.html

**Comment by ilpianista** [comment]
@Lekensteyn thank you for the workaround!
However, there's no need to bind /data/cacerts, just execute
```adb shell mount -o remount,rw /system```
copy the certificate and remount again with
```adb shell mount -o remount,ro /system```

**Comment by Lekensteyn** [comment]
@ilpianista Unfortunately that does not work on all systems. In my case I was using Android-X86 where /system is a ro CD image. Another case where it could (presumably) have bad side-effects is when "verified boot" is enabled, simply mounting a filesystem would change the filesystem metadata and fail the verification.

**Comment by ilpianista** [comment]
@Lekensteyn ah, I see. Thank you a lot for explaining.

**Comment by victorhooi** [comment]
If you have root access, this blog post might help:

https://blog.jeroenhd.nl/article/android-7-nougat-and-certificate-authorities

**Comment by fiery-** [comment]
It seems that [this research lab](https://blog.nviso.be/2017/12/22/intercepting-https-traffic-from-apps-on-android-7-using-magisk-burp/) has released a [module](https://github.com/NVISO-BE/MagiskTrustUserCerts) that is able to install client certificates on Android Nougat. There is even a screenshot with mitmproxy cert installed. Module is explained as: `Universal Systemless Interface, to create an altered mask of the system without changing the system itself. `

Did someone test this solution or it also requires a rooted device (or recompiled apps)? Removing client cert after testing is essential as well.

**Comment by Lekensteyn** [comment]
@fiery- That module seems to use Magisk which is used to root phones. Assuming it does not write directly to `/system`, I guess that it works using bind mounts as described in https://github.com/mitmproxy/mitmproxy/issues/2054#issuecomment-327735569

**Comment by fiery-** [comment]
@Lekensteyn Indeed, they use Magisk to bind-mount a user certs folder to the /system. However, installing Magisk seems to be a task for very advanced users. 

Magisk's developer posted on Twitter that there is a fake site and a fake MagiskManager app on Google Play Store:

> ![fake MagiskManager. com site](https://user-images.githubusercontent.com/4441889/35769817-b01d711e-0919-11e8-974c-2c6d71830660.png) ![fake MagiskManager app](https://user-images.githubusercontent.com/4441889/35769849-4d282c7e-091a-11e8-88e2-cb5c1c8d1cb0.png)

The [official Github](https://github.com/topjohnwu/Magisk) repository hosts only source and is not helpful how to install/uninstall this module (on non rooted devices). Some devices (i.e. Samsung) detect when you root them, so the risk to get it wrong is not small.

**Comment by hanassi** [comment]
@Lekensteyn 
I've try to place "burp suite CA" at /etc/security/cacerts at my device, and I made it system certificate. but proxy still not working. In your way, It does not work in "burp suite CA"?

**Comment by Lekensteyn** [comment]
@hanassi Have you followed the exact steps as I described [above][1], including making sure that the fingerprint is used as filename? What does "not working" mean? If you connect to https://example.com while your MITM is active, do you see your CA?

Have you looked with Wireshark what traffic (and certificate) is exchanged?

 [1]: https://github.com/mitmproxy/mitmproxy/issues/2054#issuecomment-327735569

**Comment by hanassi** [comment]
@Lekensteyn yes, I download "burp suite CA" and rename certificate by use openssl command.

openssl x509 -inform der -outform pem -out cacert.crt -in cacert.der
openssl x509 -noout -subject_hash_old -in cacert.crt

![1](https://user-images.githubusercontent.com/34124760/39664707-360ef5a0-50c2-11e8-9bf9-a46669641498.PNG)
![2](https://user-images.githubusercontent.com/34124760/39664708-3822c6e6-50c2-11e8-9676-12ec21ff0a68.PNG)
![3](https://user-images.githubusercontent.com/34124760/39664709-39730826-50c2-11e8-8911-18e17b73498c.PNG)
![4](https://user-images.githubusercontent.com/34124760/39664710-3b47faf8-50c2-11e8-9060-f0d905ecd58d.PNG)
![5](https://user-images.githubusercontent.com/34124760/39664711-3c84e854-50c2-11e8-9c3a-61dd47144d70.jpg)
![screenshot_20180506-002244](https://user-images.githubusercontent.com/34124760/39664807-54194fd0-50c4-11e8-8be2-45cbf911d261.png)
![screenshot_20180506-002403](https://user-images.githubusercontent.com/34124760/39664808-561ac28c-50c4-11e8-8644-3f38a7fe8340.png)

**Comment by pohmelie** [comment]
I also have problems with android nougat. I made everything as @Lekensteyn said, except that I can remount `/system`, so I see `mitm proxy` in list of trusted system credentials. But I'm still do not see some apps traffic through mitm proxy. In wireshark I see some handshaking and 

`1590	10.188781010	192.168.1.33	192.168.1.42	TLSv1.2	73	Alert (Level: Fatal, Description: Unknown CA)`

And tcp `rst` after.

Do not really now, maybe default mitm certificate is blacklisted or some anti-mitm technics in apps twitter, whatsapp, instagram, dropbox, youtube, uber are done. But reddit, google maps, google mail works just fine and I see all requests.

**Comment by BookGin** [comment]
I also have problems with Chrome 64 on my Android 7.0 mobile phone. I have installed the user-added mitmproxy CA, but Chrome cannot connect to any website, even in HTTP. It seems that Chrome refuses to connect if the CA certificate is not system-trusted. Here is the log:

```
192.168.1.1:46382: clientconnect
192.168.1.1:46382: CONNECT safebrowsing.googleapis.com:443
 << Cannot establish TLS with client (sni: safebrowsing.googleapis.com): TlsException("SSL handshake error: Error([('SSL routines', 'ssl3_read_bytes', 'sslv3 alert certificate unknown')],)",)
192.168.1.1:46382: clientdisconnect
``` 

However, to my surprise, Firefox 61.0.2 works very well. I can use Google search without any problem.

**Comment by RoseDeSable** [comment]
You can intercept all the traffic of your phone also such of the apps. From android 4 upward you must go the way over the sd card. 

1. Insert the sd card in the laptop or other computer

2. store the certificate mitmproxy-ca-cert.cer of mitmproxy on the sd card

3. insert the card in the tablet or phone

4. Now click on the certificate for installing

After this steps you are able to intercept with mitmproxy. It could be, that you additionally must install some root certificates for mobile devices under the cert-directory for mitmproxy

Obviously android up version up seems to accept the sd card as a trustable location, so that you mustn't root the device for installing the certificate under the directory for system certificates. The certificate is stored under user certificate, but android accept it as system certificate !

**Comment by mhils** [comment]
@RoseDeSable: As discussed earlier in the thread this is incorrect.

**Comment by RoseDeSable** [comment]
Yes, I went wrong, because only the App "AEM" correctly works. It only seems to test the certificat for iwebgis.com in the chain of the certificates. Correctly it is to control the whole chain up to the "Root Certificate" for the instance, that certifies all the followers in the chain, because in android only the "Root Certificates" are stored. Correct ?

**Comment by wych42** [comment]
@Lekensteyn I got this error after bind --mount in APP: "trust anchor for certification path not found"

**Comment by adipascu** [comment]
Installed the certificate as a system cert in Android 7, it still did not fix the issue. I think the mitmproxy certificate is lacking fields and causing issues.

**Comment by Jonahss** [comment]
@Lekensteyn I followed your instructions, but for some reason, my emulator claims it doesn't have permission to read the new certificate file.
I've checked the permissions:

```
~> adb shell ls -la /etc/security/cacerts | grep c8750f0d
-rw-r--r-- 1 root root 1318 2019-03-20 11:29 c8750f0d.0
```

And here's logcat complaining about not being able to read the certificate file:
```
03-20 12:33:52.678  4641  5106 E DirectoryCertificateSrc: Failed to read certificate from c8750f0d.0
03-20 12:33:52.678  4641  5106 E DirectoryCertificateSrc: java.io.FileNotFoundException: /system/etc/security/cacerts/c8750f0d.0 (Permission denied)
03-20 12:33:52.678  4641  5106 E DirectoryCertificateSrc:       at java.io.FileInputStream.open0(Native Method)
```

Any idea what else I can do?

**Comment by Jonahss** [comment]
Answer to my own question^

I noticed that not only was my new cert not readable, none of the default system certs were either. They all disappeared from the list in Settings.

What I needed to do for my emulator was `adb remount`. But this was blocked by all the methods I saw online except for starting the emulator with the `-writeable-system` flag. Now I'm mitm'ing :D

**Comment by Keshvadi** [comment]
I read all of the comments, but what is the conclusion. Is there any way to intercept apps traffic?

**Comment by Jonahss** [comment]
@Keshvadi totally possible. I wrote a blog post about it: https://appiumpro.com/editions/63

It contains instructions on how to set up the emulators. You can run mitmproxy from commandline rather than the library I write about and you can ignore Appium stuff if you're not testing.

All I did was use lekenstyen's answer, but had to add a flag when starting the emulator.
https://github.com/mitmproxy/mitmproxy/issues/2054#issuecomment-327735569

**Comment by KaKi87** [comment]
Hello,

[This tutorial](https://gist.github.com/pwlin/8a0d01e6428b7a96e2eb) aso works : install certificates the classic way then move the files from `/data/misc/user/0/cacerts-added` to `/system/etc/security/cacerts`.

++

**Comment by RoseDeSable** [comment]
That's correct. But I have a Samsung Tablet with version 8.1.0 of android. Here I it is impossible to mount the system partition. Rooting the tablet is not possible or too risky. 
  But it is interesting to see, that it is possible to intercept some apps with a root-certificate in the local user storage. It is dependent on the manner, how the apps are wrote. The traffic of apps like "aem" or "bmw connected drive" can be intercepted, but apps like "katwarn", "youtube" aren't breakable. 
    Some apps are using the browser of the tablet for login. Because the browser is interceptable with a certificate in the user storage, you can capture the user's password. Therefore I don't understand, why the programmers of the cloud apps - nextcloud or owncloud - use this insecure way for authentication. 
   On the other side after md5 the algorithm sha1 is also weak because of the increasing cpu and/or gpu power: collisions are quicker found as the years ago. So someone could copy the signature of a server certificate, that is signed by an authourized instance in the internet, to its own self signed root-certificate. This is possible in the case, where the hash of its own certificate is equal to the hash of the server certificate (collosion). I believe, that in this case the interception of neraly whole the traffic of a phone is possible with mitmproxy.

**Comment by Junyan** [comment]
It's working for me.
- Install VirtualXposed APP and JustTrustMe module.
- Install your APP in VirtualXposed.
- Open your APP in VirtualXposed.

references:
- https://github.com/android-hacker/VirtualXposed
- https://github.com/Fuzion24/JustTrustMe

**Comment by Jonahss** [comment]
Thanks @Junyan !

**Comment by gah-bo** [comment]
> It's working for me.

I'm trying to use VirtualXposed APP and JustTrustMe so that I can do this with WeedMaps to extract menu data from my dispensary, inspired by https://www.youtube.com/watch?v=Fa1zlLEGPtw .

mitimproxy keeps disconnecting/connecting on me and apps like WM just wont work. However i tried this VirtualXposed APP + JustTrustMe combo and when I open WM it gives me a prompt to update play services. I tried this https://repo.xposed.info/module/com.aviraxp.nomoregoogleplayservices to no avail. I cannot enter WM on VirtualXposed APP because that prompt stops me.

 I am not rooted btw.

**Comment by RoseDeSable** [comment]
Hello,
 I find this page https://virtualxposed.com/ in the internet. If I downloaded the newest version of the app from the page, I see the reference to github: https://github.com/android-hacker/VirtualXposed/releases/download/0.18.2/VirtualXposed_0.18.2.apk. Exist a reason to build the app from source (https://github.com/android-hacker/VirtualXposed) with apktool, or can I use the downloaded apk for testing ?

Bye
Rose

**Comment by RoseDeSable** [comment]
Hello,
 VirtualXposed functions in my case: Samsung Galaxy Tab A6 with android 8.10. (Samsung ExperienceVesrion 9.5). I can decrypt some apps in the container with mitmproxy. But it is too circuitous to install the app and a part of the google environment there. 
  Has someone experience with "Hooking with Frida & Objection" ? The intension is to directly click out the certificate pinning of  an app on the android system.

**Comment by Mattwmaster58** [comment]
I _almost_ had mitmproxy working for Android, but the cert generated by mitmproxy expired 2.5y in the future, [which some random guy on the internet thinks means Android disregards it](https://stackoverflow.com/a/61637667/). In my own testing this is certainly the case.

**Comment by iuriatan** [comment]
I just would like to comment that [this](https://community.tealiumiq.com/t5/Tealium-for-Android/Setting-up-Charles-to-Proxy-your-Android-Device/ta-p/5121#toc-hId-735733144) tutorial on _additional configuration steps for andoid N and above_ helped me opt in the mitm proxy for the app I'm working on. A simple change in the app manifest can override the nougat restrictions. I strongly believe it deserves mentioning in mitmproxy docs!

**Comment by kpj** [comment]
In case someone is looking for a solution which does not require modifying the APK's manifest, I believe the trick is to:

*  use API level 28 with Google APIs (*not* Google Play)
* start the emulator with `-writable-system`
* run `adb root && adb remount`

This is basically a combination a various previous posts in this thread (and is probably also not future-proof), but should allow you to install the required system certificates without issues.
I did a slightly more detailed write-up [here](https://kpj.github.io/misc/InterceptingHTTPTraffic.html).

<details>
  <summary>Interception Example</summary>
  
![image](https://user-images.githubusercontent.com/645279/94815337-b5b24a80-03fa-11eb-8c53-fd482efa5d54.png)
  
</details>

Hope this is of help to some :)

**Comment by sinhpn92** [comment]
Have any suggest to sniff traffic from Android app on Android 8+?

**Comment by RoseDeSable** [comment]
For sniffing the traffic of a specific app the app "SandroProxy" is a first good solution. In the column "APPS" you see all the traffic seperated for the single apps. You can install a self-signed root certificate for intercepting ssl-traffic. But also here: You must test, for which app you can do the interception. Apps like Youtube only accept certificate of the system storage. With an outside proxy like mitmproxy you are not able to exactly separate the traffic for a specific app !

**Comment by thijs-qv** [comment]
I was able to use mitmproxy to inspect app traffic on an emulator running Android 29. I followed the steps of @KaKi87, but had to do some extra things to make `/system` writable.

- create an emulator with Google API (not Google Play).
- start the emulator with the `-writable-system` option
- enable the proxy by setting it in the emulator settings via Settings > Proxy
- install the mitm certificate (via mitm.it) on the emulator
- Now run the following:
  `adb root`
  `adb shell avbctl disable-verification`
  `adb reboot`
 - When the emulator has rebooted we can make the `/system` directory writable:
   `adb root`
   `adb remount`
  - `/system` is now writable, so copy the user certificate from `/data/misc/user/0/cacerts-added` to `/system/etc/security/cacerts`

**Comment by aptarmy** [comment]
If you don't want to root your physical mobile device like me, you can add [Network Security Configuration](https://developer.android.com/training/articles/security-config) to the apk file as mentioned by @victorhooi
[here](https://github.com/mitmproxy/mitmproxy/issues/2054#issuecomment-289206209), but it only works with Android below 11. For Android 11, If you use [AddSecurityExceptionAndroid](https://github.com/levyitay/AddSecurityExceptionAndroid) you won't be able to install the modified apk, as your device won't accept older version of apk singing scheme.

So for Android 11, you can use somthing else like [apk-mitm](https://github.com/shroudedcode/apk-mitm). It works great with my mobile device.

install it using
`npm install -g apk-mitm`

**Comment by AtitBimali** [comment]
hey @aptarmy I tried few apps with `apk-mitm`, it successfully rebuilds the apk and the apk also gets install on my phone without any issue. But when i launch the app, it closes immediately.

**Comment by snimavat** [comment]
On Android 13 emulator : configured to use proxy (from android studio)

Android App or chrome traffic just does not go through the proxy, it behaves like there's no proxy) 
but android studio traffic does go through proxy and i can see it in logs. 

Any idea, it feels like avd is just ignoring the proxy settings.

**Comment by s1nistr4** [comment]
This is still an issue in 2023, with 55+ likes on the issue report. Please fix this issue, it should be top priority. This is the only thing keeping a lot of other users and I from using mitmproxy for everything.

**Comment by 3052** [comment]
I think this issue should be closed. I support the basic premise, but the original issue as posted is too vague to be actionable. we need to know WHAT app is having trouble, as the solution is highly dependent on what if any protections the app is using, some of which might be out of scope for MITM Proxy.

**Comment by KaKi87** [comment]
> I think this issue should be closed.

Agreed.

> the original issue as posted is too vague to be actionable. we need to know WHAT app is having trouble, as the solution is highly dependent on what if any protections the app is using, some of which might be out of scope for MITM Proxy.

Not for that reason, though.

On the contrary, the issue isn't app-specific and was clearly idenfied as the SSL certificate trust issue, for which several people (including myself) shared a fix, despite requiring root. Case solved.

In the meantime, I'm still waiting on #4408 and #5432.

**Comment by okridgway** [comment]
Hi, could someone please provide more detailed instructions on the steps you need to take to get it to work? I have mitmproxy set up no problem, but can't figure out how to get the certs to be accepted by the phone.

**Comment by RoseDeSable** [comment]
Hi,
 if mitmproxy starts the first time, it creates the directory ".mitmproxy" in your home directory with all its certificates. More under https://docs.mitmproxy.org/stable/concepts-certificates/. You can import the certificate for android "mitmproxy-ca-cert.cer" in the certificate storage for your android device. If the device is unrooted the import goes into the certificate storage for user certificates. Here only some apps accept it. You must try it. For example the browsers does it. But other apps like youtube doen't. If the device is rooted, than you must copy the certificate in the storage for system certificates. Some more apps could accept it, but nogt all !

**Comment by s1nistr4** [comment]
This has pretty much been solved, I haven't had an issue with it in newer versions, just as long as you disable ssl pinning with something like objection, it works great.

**Comment by JorySeverijnse** [comment]
What do you mean it has been solved, with a relatively new android version/phone this is not possible to do because the /system partition is locked out.
I got my certificate in the correct directory but when rebooting the system my phone will mount it as follows:
`/dev/block/dm-44 /system/etc/security/cacerts/c8750f0d.0 f2fs ro,lazytime,seclabel,noatime,background_gc=on,discard,no_heap,user_xattr,inline_xattr,acl,inline_data,inline_dentry,flush_merge,extent_cache,mode=adaptive,active_logs=6,reserve_root=32768,resuid=0,resgid=1065,inlinecrypt,alloc_mode=default,fsync_mode=nobarrier 0 0
`
This is what the entries look like 
`/dev/block/dm-2 /system/etc/security/cacerts/c7f1359b.0 ext4 ro,seclabel,relatime,discard 0 0
`
Now i have not seen a way to actually mount this partition under /dev/block/dm-2.

Also changing fstab is read only  /vendor/etc/fstab.default. Im not sure wether my phone gets bricked if i could actually change this entry to rw
`system                                                  /system                ext4    ro,barrier=1,discard                                 wait,slotselect,avb=vbmeta_system,logical,first_stage_mount,avb_keys=/avb/q-gsi.avbpubkey:/avb/r-gsi.avbpubkey:/avb/s-gsi.avbpubkey
`
Could someone let me know if they have found a workaround. Im using lineageos 21 on a rooted oneplus 9 pro

**Comment by yonatan1811** [comment]
Hi, I've been trying to install mitmproxy cert and i could do so. I could see it in the root trusted certificates as mitmproxy. I couldn't go into any website without privacy error. Just when I installed the certificate manually using the <install from sd card> the emulator did trusted the certificate and i could navigate to any website i want. But I couldn't see apps traffic. every time there is a reqeuest going out i have a warning message that the client does'nt trust the tls certificate. so i tried using burp cert and it did accually work, i can access any (i think (: ) app/website i want and see it's trafiic.
iv'e downloaded the certificate using http://mitm.it. 
i'm using ldplayer 9 and emulator android 9.0.
Thank you in advanced. (the burp certificate was downloaded automatically using adb push... there isn't a certificate on the user trusted certificates).

**Comment by RoseDeSable** [comment]
Hi,
  it sees, as if the cerificate is stored in the "devices trust storage". This location is on a security level, where all apps trust the certificates, that are stored there. The reason: this location is part of the system ! So you can intercept the traffic of all your apps with burp. Hurra !
 The question is: What have you or burp's installation procedure done on your android 9 emulation, to receive root rights for writing in this location of the system

**Comment by yonatan1811** [comment]
> Hi, it sees, as if the cerificate is stored in the "devices trust storage". This location is on a security level, where all apps trust the certificates, that are stored there. The reason: this location is part of the system ! So you can intercept the traffic of all your apps with burp. Hurra ! The question is: What have you or burp's installation procedure done on your android 9 emulation, to receive root rights for writing in this location of the system

Well, I don't remember exactly what I have done, but i'm sure that i have not used any external module to install burp's certificate on the system trusted certs. The way i fixed it with mitmproxy cert is I used a magisk module called "AlwaysTrustUserCerts". this way after granting root with magisk you could install the mitm cert and then install the module, restart the emulator , and there you have it. the mitm certificate is installed in the right place and you could see both application\web traffic.
Have a good one!

**Comment by RoseDeSable** [comment]
Yes,
 you can use magisk on your emulation. But the world of a real android device is a another. For example, Samsung protects the newer devices with a security circuit. If you root the device by installing the magisk partition, the firmware notice this and let burn the circuit. Now the device is unusable.

**Comment by jnlow366** [comment]
Is there just no fix there?  Samsung S24Ultra isn't something you can root or change system partitions.  Short of risking bricking the phone, is there no way to install an intended VPN+Proxy?

**Comment by RoseDeSable** [comment]
Hi,
 I don't believe, that kids apps are able to intercept the ssl-traffic. If it would be so, hackers would decompile the app for copying the interception pieces of the code in their attacking programs. That would be a very great weakness, because whole the world would suddenly sit in a glass house. I believe, that the kids apps are only noticing the traffic in the manner of tcpdump.  
 If you use vpn-products like openvpn, wireguard, checkpoint's endpoint security, ... you always have an over-encryption. I mean, that the apps like youtube, facebook, Germany's Katwarn or NINA, ... are encyrypt their traffic and the vpn app additionally encrypt this traffic in its own ssl-tunnel. Wireguard is an app on your samsung device. If it accept mitmproxy's certificate in the user's certificate store of your device, mitmproxy is able to intercept the traffic: but only  the traffic of the wireguard tunnel   n o t   the ssl-traffic of the client apps like youtube. You only see like in wireshark the encrypted traffic of the client-apps.
